syntax = "proto3";

package faiss;

service FaissService {
    rpc initFaiss (initFaissRequest) returns (initFaissResponse);
    rpc addVectors (addVecRequest) returns (addVecResponse);
    rpc deleteVectors (deleteVecRequest) returns (deleteVecResponse);
    rpc getNearest (getNearestRequest) returns (getNearestResponse);
}

message vectorType {
    repeated float e = 1;
}

message documentType {
    string _id = 1;
    vectorType vector = 2;
}

message initFaissRequest {
    int32 nlist = 1;
    int32 nprobe = 2;
    int32 bpv = 3;
    int32 bpsv = 4;
    int32 vd = 5;
    repeated vectorType matrix = 6;
}

message initFaissResponse {
    bool status = 1;
}

message addVecRequest {
    repeated documentType documents = 1;
}

message addVecResponse {
    bool status = 1;
    repeated string _id = 2;
}

message deleteVecRequest {
    repeated string _id = 2;
}

message deleteVecResponse {
    bool status = 1;
    repeated string _id = 2;
}

message getNearestRequest {
    repeated vectorType matrix = 1;
    int32 k = 2;
}

message getNearestResponse {
    bool status = 1;
    repeated string _id = 2;
    string matrix = 3;
}